File: frontend/lib/backtest.ts
import { isoDate } from "./db";
import type { sql as sqlFn } from "./db";

function mom(current: number | null, previous: number | null): number | null {
  if (current == null || previous == null || previous === 0) return null;
  return Math.round((1000 * (current - previous)) / previous) / 10;
}

function corr(xs: number[], ys: number[]): number | null {
  const n = xs.length;
  if (n < 2) return null;
  const mx = xs.reduce((a, b) => a + b, 0) / n;
  const my = ys.reduce((a, b) => a + b, 0) / n;
  let num = 0;
  let dx = 0;
  let dy = 0;
  for (let i = 0; i < n; i++) {
    num += (xs[i] - mx) * (ys[i] - my);
    dx += (xs[i] - mx) ** 2;
    dy += (ys[i] - my) ** 2;
  }
  if (dx === 0 || dy === 0) return null;
  return Math.round((num / Math.sqrt(dx * dy)) * 10000) / 10000;
}

export async function computeBacktest(q: ReturnType<typeof sqlFn>) {
  const apix = (await q`
    SELECT period_date, value FROM index_values
    WHERE series = 'apix_laspeyres' AND frequency = 'monthly' AND origin IS NULL
  `) as { period_date: string; value: number }[];
  const apixBy = new Map(apix.map((r) => [isoDate(r.period_date), r.value]));

  const tmu = (await q`
    SELECT month, value FROM dgca_benchmark WHERE metric = 'tmu_basket_index'
  `) as { month: string; value: number }[];
  const tmuBy = new Map(tmu.map((r) => [isoDate(r.month), r.value]));

  const quotes = (await q`
    SELECT origin, destination, collected_on, total_fare FROM quotes_clean WHERE is_outlier = 0
  `) as { origin: string; destination: string; collected_on: string; total_fare: number }[];
  const routeAvg = new Map<string, number[]>();
  for (const qrow of quotes) {
    const d = isoDate(qrow.collected_on);
    const month = `${d.slice(0, 7)}-01`;
    const key = `${month}|${qrow.origin}|${qrow.destination}`;
    const list = routeAvg.get(key) ?? [];
    list.push(qrow.total_fare);
    routeAvg.set(key, list);
  }

  const dgcaRoute = (await q`
    SELECT month, origin, destination, value FROM dgca_benchmark WHERE metric = 'tmu_route_avg_fare'
  `) as { month: string; origin: string; destination: string; value: number }[];
  const dgcaRouteMap = new Map(
    dgcaRoute.map((r) => [`${isoDate(r.month)}|${r.origin}|${r.destination}`, r.value]),
  );

  const months = [...new Set([...apixBy.keys(), ...tmuBy.keys()])].sort();
  const rows: Record<string, unknown>[] = [];
  let prevApix: number | null = null;
  let prevTmu: number | null = null;
  for (const month of months) {
    const a = apixBy.get(month) ?? null;
    const t = tmuBy.get(month) ?? null;
    rows.push({
      month,
      apix_monthly: a,
      apix_mom_pct: a != null ? mom(a, prevApix) : null,
      dgca_value: t,
      dgca_mom_pct: t != null ? mom(t, prevTmu) : null,
    });
    if (a != null) prevApix = a;
    if (t != null) prevTmu = t;
  }

  for (const [key, fares] of [...routeAvg.entries()].sort()) {
    const [month, origin, dest] = key.split("|");
    const avg = fares.reduce((a, b) => a + b, 0) / fares.length;
    rows.push({
      month,
      origin,
      destination: dest,
      route_avg_fare: Math.round(avg * 100) / 100,
      dgca_route_avg: dgcaRouteMap.get(key) ?? null,
    });
  }

  const published = (await q`
    SELECT month, value FROM dgca_benchmark WHERE metric = 'tmu_72_route_index'
  `) as { month: string; value: number }[];
  for (const r of published) {
    rows.push({ month: isoDate(r.month), dgca_value: r.value });
  }

  const paired = rows.filter(
    (r) => r.origin == null && r.apix_monthly != null && r.dgca_value != null,
  ) as { apix_monthly: number; dgca_value: number }[];

  return {
    correlation: corr(
      paired.map((r) => r.apix_monthly),
      paired.map((r) => r.dgca_value),
    ),
    n_pairs: paired.length,
    note: "Monthly APIx vs published DGCA TMU 72-route composite. Quotes live in Neon.",
    rows,
  };
}

File: backend/app/services/backtest.py
from __future__ import annotations

from collections import defaultdict
from datetime import date

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.models import DgcaBenchmark, IndexValue, QuoteClean
from app.schemas import BacktestRow, BacktestSummary


def _mom(current: float | None, previous: float | None) -> float | None:
    if current is None or previous is None or previous == 0:
        return None
    return round(100.0 * (current - previous) / previous, 2)


def _corr(xs: list[float], ys: list[float]) -> float | None:
    n = len(xs)
    if n < 2:
        return None
    mx = sum(xs) / n
    my = sum(ys) / n
    num = sum((x - mx) * (y - my) for x, y in zip(xs, ys))
    dx = sum((x - mx) ** 2 for x in xs) ** 0.5
    dy = sum((y - my) ** 2 for y in ys) ** 0.5
    if dx == 0 or dy == 0:
        return None
    return round(num / (dx * dy), 4)


def compute_backtest(session: Session) -> BacktestSummary:
    apix = session.scalars(
        select(IndexValue).where(
            IndexValue.series == "apix_laspeyres",
            IndexValue.frequency == "monthly",
            IndexValue.origin.is_(None),
        )
    ).all()
    apix_by_month = {r.period_date: r.value for r in apix}

    tmu = session.scalars(
        select(DgcaBenchmark).where(DgcaBenchmark.metric == "tmu_basket_index")
    ).all()
    tmu_by_month = {r.month: r.value for r in tmu}

    route_apix_avg: dict[tuple[date, str, str], list[float]] = defaultdict(list)
    quotes = session.scalars(select(QuoteClean).where(QuoteClean.is_outlier == 0)).all()
    for q in quotes:
        month = date(q.collected_on.year, q.collected_on.month, 1)
        route_apix_avg[(month, q.origin, q.destination)].append(q.total_fare)

    dgca_route = session.scalars(
        select(DgcaBenchmark).where(DgcaBenchmark.metric == "tmu_route_avg_fare")
    ).all()
    dgca_route_map = {(r.month, r.origin, r.destination): r.value for r in dgca_route}

    months = sorted(set(apix_by_month) | set(tmu_by_month))
    rows: list[BacktestRow] = []
    prev_apix = None
    prev_tmu = None
    xs: list[float] = []
    ys: list[float] = []
    for month in months:
        a = apix_by_month.get(month)
        t = tmu_by_month.get(month)
        rows.append(
            BacktestRow(
                month=month,
                apix_monthly=a,
                apix_mom_pct=_mom(a, prev_apix) if a is not None else None,
                dgca_value=t,
                dgca_mom_pct=_mom(t, prev_tmu) if t is not None else None,
            )
        )
        if a is not None:
            prev_apix = a
        if t is not None:
            prev_tmu = t
        if a is not None and t is not None:
            xs.append(a)
            ys.append(t)

    for (month, origin, dest), fares in sorted(route_apix_avg.items()):
        dg = dgca_route_map.get((month, origin, dest))
        avg = sum(fares) / len(fares)
        rows.append(
            BacktestRow(
                month=month,
                origin=origin,
                destination=dest,
                route_avg_fare=round(avg, 2),
                dgca_route_avg=dg,
            )
        )
        if dg is not None:
            xs.append(avg)
            ys.append(dg)

    published = session.scalars(
        select(DgcaBenchmark).where(DgcaBenchmark.metric == "tmu_72_route_index")
    ).all()
    for r in published:
        rows.append(
            BacktestRow(
                month=r.month,
                dgca_value=r.value,
            )
        )

    return BacktestSummary(
        correlation=_corr(
            [row.apix_monthly for row in rows if row.apix_monthly is not None and row.dgca_value is not None and row.origin is None],
            [row.dgca_value for row in rows if row.apix_monthly is not None and row.dgca_value is not None and row.origin is None],
        ),
        n_pairs=sum(
            1
            for row in rows
            if row.origin is None and row.apix_monthly is not None and row.dgca_value is not None
        ),
        note=(
            "Monthly APIx (Laspeyres, lowest-economy Jevons elementary) vs TMU-style mean fares "
            "on the same reconstructed sample, plus published 72-route TMU composite (+20.5% Jun 2026 vs Mar 2025). "
            "DGCA does not publish a public daily route-level fare dump; overlapping months are the validation set."
        ),
        rows=rows,
    )
